### ISSUE TITLE
[Feature]: Integrate Writing in the Margins inference pattern ($5,000 Bounty)

### ISSUE THREAD

**Issue by melisa-writer** [vllm-project/vllm#9807]
### 🚀 The feature, motivation and pitch

[Writer](www.writer.com) has introduced ["Writing in the Margins" algorithm (WiM)](https://arxiv.org/html/2410.05258v1) that 
boosts results for long context window retrieval. The task is composed from "context" and "query" that is put at the end.

The basic idea is to generate additional text while doing chunked prefill. The extra decoding step does not contribute to the KV-cache prefilling. The text is later concatenated and incorporated into the final answer just before the query.

There exists a pure HuggingFace transformers implementation: [https://github.com/writer/writing-in-the-margins](https://github.com/writer/writing-in-the-margins)

This is a high level overview of the inference pattern:
<img width="701" alt="Screenshot 2024-10-29 at 18 33 22" src="https://github.com/user-attachments/assets/52b1d1ed-3232-4024-9f3a-032259606d24">

And this is more detailed explanation how to do it efficiently by batch generation and prefill requests.
<img width="352" alt="Screenshot 2024-10-29 at 18 34 19" src="https://github.com/user-attachments/assets/69bc2fdb-e750-4298-a1ec-7f8554f75100">

The algorithm itself:
<img width="361" alt="Screenshot 2024-10-29 at 18 39 05" src="https://github.com/user-attachments/assets/ca5b79a9-0aa0-41ee-9477-1ba7d78cd3fc">

The expected solution can be a feature added to vllm or a vllm fork, we are happy to maintain it. 
The WiM solution assumes extra input preprocessing steps (nltk splitting) and variable chunk size for chunked prefill, but those details can be left out from the solution.

We offer $5,000 bounty for the main contributor (but the bounty can be shared if there is more than one developer involved).

paper: [ArXiv](https://arxiv.org/html/2410.05258v1)
press coverage: 
- [#1 Paper of the day in HuggingFace](https://huggingface.co/papers?date=2024-08-28)
- [Paper Club](https://www.youtube.com/watch?v=VHwrhL_MSV4&ab_channel=LatentSpace)
- [Writer's blog](https://writer.com/engineering/writing-in-the-margins/)

### Additional context

Github: [https://github.com/writer/writing-in-the-margins](https://github.com/writer/writing-in-the-margins)

### Before submitting a new issue...

- [X] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.

**Comment by noooop** [comment]
That's a brilliant idea. 


I think

This algorithm can be implemented efficiently through prefix cache, and vllm supports prefix cache.

So you can implement this algorithm without modifying any vllm code.

specific

1. start vllm with prefix cache open
2. submit Margin generation requests in order
3. immediately submit the next Margin generation request when the previous request returns the first token.


How do I get $5,000? （joke）

**Comment by melisa-writer** [comment]
Thank you!

Indeed Automatic Prefix Caching could be used to simulate WiM algorithm. However there are a few issues related with this:

1. Different tokenization: 

The real issue is the different tokenization you get by sending the text "A gentle breeze stirred" and then "the leaves as children" or the text "A gentle breeze stirred the leaves as children". To really apply WiM by exploiting Prefix Caching, you need to send multiple requests trimmed at the exact point in which you want the tokenizer to be break the text, but this would mean sending multiple requests to vLLM (which is what you do when you "simulate it"), which would result in much longer time to process compared to prefill-generate-prefill-generate done in the same request.

2. High workload KV cache eviction:  

In a high workload scenario, the KV-Cache would be evicted, and in case of cache-miss it would be re-created again. But is it a problem? It may be. The whole point of WiM is to reuse the partially prefilled KV-Cache (or in other terms, KV prefixes).

3. Storing customer data: 
Automatic prefix caching means user data is stored in RAM. We would rather turn that feature off due to compliance issues.

To summarise: Good point with Automatic Prefix Caching, it can be used for prototyping. We still need something different for production use case.

**Comment by noooop** [comment]
Use prompt_token_ids as input can bypass the tokenizer. 

The previous Margin generation is still in progress, so KV-Cache is still on the GPU. You cannot miss it.

Prefix Caching is almost a perfect solution

**Comment by apehex** [comment]
I've just digested the reference material, that's a nice optimization! :)

From what I understand, the margin size should be limited to a small number of tokens.
Hence, their classification would be comparatively fast, right?

That said, I have a small question / suggestion on the extraction prompt.
Would it be possible to both generate and classify the margins at once?
With a prompt similar to:

```
Extract text relevant to the query: "{query}".
Write everything you find in your answer, without anything else.
If there is no related data in the context above, write "NOTHING FOUND" instead.
```

WDYT?

**Comment by apehex** [comment]
How would you define **a satisfying extension of vLLM** to support WiM?

I think it'd  start by writing tests.
For example if Llama 3.1 misses the answer on "babilong_64k" it could be a good test.
Also, test that the retrieval performance of the model is always above the regular inference.

**Comment by whyplum** [comment]
> I've just digested the reference material, that's a nice optimization! :)
> 
> From what I understand, the margin size should be limited to a small number of tokens. Hence, their classification would be comparatively fast, right?
> 
> That said, I have a small question / suggestion on the extraction prompt. Would it be possible to both generate and classify the margins at once? With a prompt similar to:
> 
> ```
> Extract text relevant to the query: "{query}".
> Write everything you find in your answer, without anything else.
> If there is no related data in the context above, write "NOTHING FOUND" instead.
> ```
> 
> WDYT?

Yes! That is actually what we did. The first token of the extracted margin was `NO/YES`, so during the decoding step, if the first encountered token was `NO`, we would drop the generation. It would be good to allow users to customise this prompt (different models will work best with different prompting strategy) and let them setting a stop token/sequence.

<img width="352" alt="Image" src="https://github.com/user-attachments/assets/acc9afab-8a00-4aad-8684-53ed55ec2b22" />

**Comment by melisa-writer** [comment]
> I've just digested the reference material, that's a nice optimization! :)
> 
> From what I understand, the margin size should be limited to a small number of tokens. Hence, their classification would be comparatively fast, right?
> 
> That said, I have a small question / suggestion on the extraction prompt. Would it be possible to both generate and classify the margins at once? With a prompt similar to:
> 
> ```
> Extract text relevant to the query: "{query}".
> Write everything you find in your answer, without anything else.
> If there is no related data in the context above, write "NOTHING FOUND" instead.
> ```
> 
> WDYT?

Yes! That is actually what we did. The first token of the extracted margin was `NO/YES`, so during the decoding step, if the first encountered token was `NO`, we would drop the generation. It would be good to allow users to customise this prompt (different models will work best with different prompting strategy) and let them setting a stop token/sequence.

<img width="352" alt="Image" src="https://github.com/user-attachments/assets/db7cb982-20f1-471a-94dd-e1b8e2b7cdd5" />

**Comment by melisa-writer** [comment]
> How would you define **a satisfying extension of vLLM** to support WiM? [@melisa-writer](https://github.com/melisa-writer)
> 
> I think it'd be good to start with tests. For example if Llama 3.1 misses the answer on "babilong_64k" it could be a good test. Also, test that the retrieval performance of the model is always above the regular inference.

Running tests is a good idea! You can use this dataset for tests: 
https://huggingface.co/datasets/Writer/writing-in-the-margins-multihoprag

Good indicator of correct implementation would be the same result of vllm and vanilla transformer implementation (assuming deterministic inference with temperature 0). In practice it might not be always possible due to the sequence length, but maybe you could subsample shorter examples from the dataset.

**Comment by apehex** [comment]
Thanks for the detailed replies! It makes a lot of sense :)

**Comment by apehex** [comment]
> Yes! That is actually what we did. The first token of the extracted margin was `NO/YES`, so during the decoding step, if the first encountered token was `NO`, we would drop the generation. It would be good to allow users to customise this prompt (different models will work best with different prompting strategy) and let them setting a stop token/sequence.
> 
> <img alt="Image" width="352" src="https://private-user-images.githubusercontent.com/52930996/417063375-db7cb982-20f1-471a-94dd-e1b8e2b7cdd5.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDA1Njk5NjgsIm5iZiI6MTc0MDU2OTY2OCwicGF0aCI6Ii81MjkzMDk5Ni80MTcwNjMzNzUtZGI3Y2I5ODItMjBmMS00NzFhLTk0ZGQtZTFiOGUyYjdjZGQ1LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAyMjYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMjI2VDExMzQyOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTFhZTBiNDdlYzU5MTFjZWE3ZGQ2MGQ3Y2E4YzM4ZWQ1MmRiOGI3ZDJmYzk1NWNlOTdmZDE2ODQ0ZjZjMzI4NzMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.825Via-2Aeeph-uk88MtCPztzlYTx-jGxZ0Jr0bg5ZI">

Great, we won't have to use KV cache tricks for this :D

Indeed the prompt will be key, I'll add a parameter for the template.
I'm not sure yet if I'll distinguish the extraction part from the classification part.
The issue will be to identify the model's decision when the classification label is set by the user (I chose "NOTHING FOUND" and you went for "YES#" and "NO#").
I'll think of something and let you know!

**Comment by apehex** [comment]
> Why do we use them? Because as you can see in the [following lines](https://github.com/writer/writing-in-the-margins/blob/ba79a5ce7c904305f423bbf3207a7924c35eef7e/run.py#L204-L207), we need to remove the generated margin and the `generation_header` from the KV-Cache to continue prefilling the next chunk of the context, so we need to know how many tokens the margin is composed of (this is easy, because we just generated it [in these lines](https://github.com/writer/writing-in-the-margins/blob/ba79a5ce7c904305f423bbf3207a7924c35eef7e/run.py#L188-L196)), but we also need to remove the `generation_header` that forces the language model to act as an assistant.
> 
> You can know how many tokens there are in the `extractive summary prompt` along with the generation header if you count the tokens in the KV-Cache before adding the `extractive summary prompt` itself, and this is exactly what we do in our reference implementation at [this line](https://github.com/writer/writing-in-the-margins/blob/ba79a5ce7c904305f423bbf3207a7924c35eef7e/run.py#L162-L167).

It might be that a few blocks of KV cache covering the `extractive summary prompt` are not reused, I'm not sure yet whether I'll have to manually clear the cache in vLLM.
I'd have to make the input chunking match the separation between context and query.

> Can you merge the two prompts? I don't think so, because you need to add the "generation header" to force the language model to generate the margin. And between the `user_header` and the `generation_header` there may be either the context the model needs to refer to or instructions to generate the margin.

Yes, I hoped to find a shortcut but I don't see how to do it!
 
> I hope it's more clear now. Let us know!

It is! Thanks a lot for the clarifications :)

**Comment by apehex** [comment]
So I'll start by writing tests.
I'm hoping to find an efficient implementation and keep the compute at a minimum 😅 

Where can I post updates for you?
There's the discord thread for example:
https://discord.com/channels/879548962464493619/1338571410838126622

**Comment by apehex** [comment]
Sorry about the silence, I took some time off!

So I've reproduced the WiM process with:
- batched requests, with the iterative logic outside vLLM
- chunked attention (2048) and prefix caching
- quantized or distilled models
- low temperature, close to deterministic

I've bundled it in a Colab notebook, I can share it if you want.

**Comment by apehex** [comment]
Also, there's the issue of isolating the query and the context, since vLLM works on fully formed prompts.

I see several paths:
1. the context is inside the input prompt and the query for the margins is specified in a separate parameter like `margin_query`
2. swap the roles and set the context in a dedicated parameter like `margin_context`
3. rely on the prompt template and parse a required / predefined section, like [`documents`][huggingface-docs-template]

Do you want me to choose on my own or should I ask you about similar decisions?

**Comment by apehex** [comment]
I wonder: is there any guarantee that the vLLM team will accept the PR?
